src/plugins_3rdparty/primer3/src/Primer3Task.cpp

```diff
@@ -574,129 +663,63 @@ void Primer3Task::selectPairsSpanningIntron(primers_t& primers, int toReturn) {
-            }
-            return;
-        } else {
-            QByteArray seq = settings.getSequence();
-            seq.append(seq.left(settings.getIncludedRegion().endPos() - settings.getSequenceSize() - settings.getFirstBaseIndex()));
-            settings.setSequence(seq);
-        }
-    }
+    const auto& sequenceRange = settings->getSequenceRange();
+    int sequenceSize = settings->getSequenceSize();
 
-    addPrimer3Subtasks(settings, regionTasks);
+    const auto& includedRegion = settings->getIncludedRegion();
+    int fbs = settings->getFirstBaseIndex();
+    int includedRegionOffset = includedRegion.startPos != 0 ? includedRegion.startPos - fbs : 0;
+    CHECK_EXT(includedRegionOffset >= 0, stateInfo.setError(tr("Incorrect summ \"Included Region Start + First Base Index\" - should be more or equal than 0")), );
 
-    if (settings.isSequenceCircular() && settings.getIncludedRegion().startPos == 1 &&
-        settings.getIncludedRegion().length == settings.getSequenceSize()) {
-        // Based on conversation with Vladimir Trifonov:
-        // Consider the start position in the center of the sequence and find primers for it.
-        // This should be enough for circular primers search.
-        QByteArray oppositeSeq = settings.getSequence().right(median);
-        oppositeSeq.append(settings.getSequence().left(settings.getSequenceSize() - median));
-        Primer3TaskSettings circSettings = settings;
-        circSettings.setSequence(oppositeSeq, true);
+    if (sequenceRange.endPos() > sequenceSize + includedRegionOffset) {
+        SAFE_POINT_EXT(settings->isSequenceCircular(), stateInfo.setError("Unexpected region, sequence should be circular"), );
 
-        addPrimer3Subtasks(circSettings, circRegionTasks);
+        QByteArray seq = settings->getSequence();
+        seq.append(seq.left(sequenceRange.endPos() - sequenceSize - fbs));
+        settings->setSequence(seq);
     }
+
+    primer3Task = new Primer3Task(settings);
+    addSubTask(primer3Task);
 }
 
 Task::ReportResult Primer3SWTask::report() {
-    for (Primer3Task* task : qAsConst(regionTasks)) {
-        bestPairs.append(task->getBestPairs());
-        singlePrimers.append(task->getSinglePrimers());
-    }
-
-    for (Primer3Task* task : qAsConst(circRegionTasks)) {
-        // Relocate primers that were found for sequence split in the center.
-        foreach (PrimerPair p, task->getBestPairs()) {
-            relocatePrimerOverMedian(p.getLeftPrimer());
-            relocatePrimerOverMedian(p.getRightPrimer());
-            if (!bestPairs.contains(p)) {
-                bestPairs.append(p);
-            }
-        }
-
-        foreach (Primer p, task->getSinglePrimers()) {
-            relocatePrimerOverMedian(&p);
-            if (!singlePrimers.contains(p)) {
-                singlePrimers.append(p);
-            }
-        }
-    }
+    CHECK(primer3Task != nullptr, Task::ReportResult_Finished);
```

Comment:
Check if the task failed before registering results.

Reply:
OK